index.js

```diff
@@ -2995,66 +3135,241 @@ function Env(handle) {
     };
 })();
 
-let _artMethodSpec = null;
-function getArtMethodSpec(vm) {
-    if (_artMethodSpec !== null)
-        return _artMethodSpec;
-
-    vm.perform(() => {
-        const env = vm.getEnv();
-        const process = env.findClass("android/os/Process");
-        const setArgV0 = env.getStaticMethodId(process, "setArgV0", "(Ljava/lang/String;)V");
+function resolveStructSpec(vm) {
+    if (this._current !== undefined)
+        return this._current;
 
-        const runtimeModule = Process.getModuleByName("libandroid_runtime.so");
-        const runtimeStart = runtimeModule.base;
-        const runtimeEnd = runtimeStart.add(runtimeModule.size);
+    let specByPointerSize = this[runtime.androidVersion.substr(0, 4)];
+    if (specByPointerSize === undefined) {
+        if (this._default) {
+            let spec = this._default(vm);
+            if (k) {
+                this._current = spec;
+                return spec;
+            }
+        }
+        specByPointerSize = this.git;
+    }
 
-        const expectedAccessFlags = kAccPublic | kAccStatic | kAccFinal | kAccNative;
+    this._current = specByPointerSize && specByPointerSize[Process.pointerSize];
+    return this._current;
+}
 
-        let jniCodeOffset = -1;
-        let accessFlagsOffset = -1;
-        let remaining = 2;
-        for (let offset = 0; offset !== 64 && remaining !== 0; offset += 4) {
-            const field = setArgV0.add(offset);
+const getArtMethodSpec = resolveStructSpec.bind({
```

Comment:
@oleavr Yup. For Android 5.0.x 32-bit the offset calculation of `entry_point_from_quick_code_` is wrong as the width of the preceding field is always 64-bit (`uint64_t`) no matter which CPU architecture the system is on. I've had no problem for 64-bit or other versions though, so I just kept the offset detection as a fallback for future versions. It may be okay to make it a default for every versions except Lollipop, but I'm a bit unsure..

Additionally, support for 64-bit Android version is missing due to insufficient offset spec info of art::Runtime and art::ClassLinker (mandatory for correct resolution of art_quick_generic_jni_stub so that stack walking would correctly recognize it's actually a JNI method when VM is in preloaded OAT image mode). We could probably work around this by injecting a native stub method and using its quickCode as *the* JNI stub. Maybe I should work on it later...